src/gotcha_dl.c

```diff
@@ -0,0 +1,36 @@
+#include "gotcha_dl.h"
+#include "tool.h"
+#include "libc_wrappers.h"
+#include "elf_ops.h"
+
+void* dlopen_wrapper(const char* filename, int flags){
+  void* handle = orig_dlopen(filename,flags);
+  filterLibrariesByName("libdl");
+  struct binding_t* tool_iter = get_bindings();
+  for(;tool_iter!=NULL;tool_iter = tool_iter->next_binding){
+    /**
+     * If you remove this comparison, you will overwrite
+     * libdl's dlsym relocation with gotcha's, leading to
+     * infinite recursion. This is cool, but inefficient.
+     */
+    if(gotcha_strcmp(tool_iter->tool->tool_name, "gotcha")!=0){
+      gotcha_wrap(tool_iter->user_binding,tool_iter->user_binding_size,tool_iter->tool->tool_name);
```

Comment:
Re: C

```
#1  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#2  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#3  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#4  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#5  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#6  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#7  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#8  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#9  0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#10 0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
#11 0x00002aaaaacd501f in dlsym_wrapper (handle=0x6041d0, symbol_name=0x400d4e "test_return_five") at /g/g0/dzpolia/src/gotcha/src/gotcha_dl.c:36
```

That goes on for at least 2000 stack frames. I'm honestly not quite sure why any more, give me a bit to reason about why this is what it is